### ISSUE TITLE
Optimize Sync All

### ISSUE THREAD

**Issue by mahaloz** [angr/binsync#137]
## Problem
`sync_all` (which is `fill_all` in the Controller), was one of the most used features at DEF CON Quals 22, however, it could be **super slow** at times. Here are some optimization ideas.

## Proposal
In the case of @zardus, he did a `sync_all` maybe every 20 mins. In this time, most functions were **not** changed, yet we still iterated over them and tried to do a `fill_function`, which is where our slowdown occurs. We need to avoid `fill_function` at all costs. We can do this by not trying to fill functions that have not changed since the last sync. How:

1. Make a dict (or class), containing each User with an index for each function. 
2. Any time a user syncs something, record the timestamp for the function index
3. Before entering `fill_function`, check if the timestamp of the function that is attempting to be synced is later than the dict stored one
4. If its not later, don't enter fill_function and report that no change needed. 

This dict system should be implemented in the common controller. The way we check for this cache can either be through a decorator or through a check actually inside fill_function.

**Comment by rhelmot** [comment]
A thing that could be an important byproduct of this refactor is that if we have the last-modified time for each function, you can have a UI element which turns yellow whenever you're working on a function that's out of date

**Comment by mahaloz** [comment]
Hehe, funny enough @rhelmot we just discussed this last BinSync meeting. It's on the board for @frqmod. The idea is to actually get rid of the `last_change` column altogether in the functions table and do color coding to tell how long ago it was changed. We then plan to put this in the native IDA function table too. 

I guess your idea is a little different actually, because you want it highlighted only if you change is less recent then someone else. 

However, I will say our end vision (as controversial as it is), may be to replace the native ida function all together. Like people will just not use the IDA one anymore. We'll see though.
